### PR TITLE
AP_PiccoloCAN: Add support for CBS can servo

### PR DESCRIPTION
**Pull request by SchrodingersGat**
This PR adds support for the CBS range of CAN servos developed by Currawong Engineering:

https://www.currawong.aero/servos-escs/can-servo/

For reference, it follows on from integration of support for the Velocity ESC - https://github.com/ArduPilot/ardupilot/pull/15425

The following functionality is provided

- Support for the servo CAN protocol
- Individual selection of servo channels to be pushed out over CAN
- Reception of telemetry data for the servos
- Logging support for servo telemetry

### CODE REVIEW

**Comment by SchrodingersGat** [comment]
@tridge @bugobliterator please let me know if you want any higher-level discussions of this PR or our CAN servo tech.

**Comment by SchrodingersGat** [comment]
I also have a couple of questions regarding implementation of "smart" servo devices. If the following questions are better addressed elsewhere, please let me know.

### Telemetry

Currently it appears there is no option for sending servo telemetry back to the GCS via Mavlink, which is possible for ESCs. Is this something that could be implemented? Or is it already implemented and I just have not been able to find it?

### Status Flags

Both the Velocity ESC and the CBS Servo have a rich set of status flags available in the telemetry data which (at the very least) would be useful to have in the flight logs. Could the logger data structures (for both the ESC and the Servo) be extended to allow "generic" status flags (as different manufacturers would have different status bits). For example it could be just a few unsigned integer fields?

**Comment by peterbarker** [comment]
> Currently it appears there is no option for sending servo telemetry back to the GCS via Mavlink, which is possible for ESCs. Is this something that could be implemented? Or is it already implemented and I just have not been able to find it?

That's the `ESC_TELEMETRY_x_TO_y` message set:

https://github.com/ardupilot/ardupilot/blob/master/libraries/AP_KDECAN/AP_KDECAN.cpp#L712

> ### Status Flags
> 
> Both the Velocity ESC and the CBS Servo have a rich set of status flags available in the telemetry data which (at the very least) would be useful to have in the flight logs. Could the logger data structures (for both the ESC and the Servo) be extended to allow "generic" status flags (as different manufacturers would have different status bits). For example it could be just a few unsigned integer fields?

I'm loathe to introduce vendor-specific-for-semantics bitmasks into the logger messages.  But there are options - for example, logging a separate message if flag changes are detected, or mapping from the vendor-specific flag set into a more generic flag set.  If you can provide a link to the set of flags you're currently thinking about I might be able to reason more...
\

File: libraries/AP_PiccoloCAN/piccolo_protocol/ServoDefines.h [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+ */
+typedef struct
+{
+    unsigned   enabled : 1;     //!< Servo enabled status
+    ServoModes mode;            //!< Servo mode
+    unsigned   cmdReceived : 1; //!< Set if the servo has received a valid position command within the configured timeout period
+    unsigned   reservedA : 1;   //!< Reserved for future use
+    unsigned   reservedB : 1;   //!< Reserved for future use
+    unsigned   reservedC : 1;   //!< Reserved for future use
+}Servo_StatusBits_t;
```
Comment (SchrodingersGat) [inline libraries/AP_PiccoloCAN/piccolo_protocol/ServoDefines.h:46]: @peterbarker for example these status bits (and the warning and error bits below)

**Comment by SchrodingersGat** [comment]
> That's the ESC_TELEMETRY_x_TO_y message 

That applies for ESC but it appears there are no downlink messages specific for servos?

> I'm loathe to introduce vendor-specific-for-semantics bitmasks into the logger messages. But there are options - for example, logging a separate message if flag changes are detected, or mapping from the vendor-specific flag set into a more generic flag set. If you can provide a link to the set of flags you're currently thinking about I might be able to reason more...

See the code I tagged you in above.

**Comment by SchrodingersGat** [comment]
@peterbarker any chance you could spare some time to have another look at this?

**Comment by bugobliterator** [comment]
@SchrodingersGat please resolve the conflict, I will review and raise this PR on the next EU DevCall.

**Review by bugobliterator — CHANGES_REQUESTED** [review]
Generally this looks good to me. There are few minor issues, I have made inline review comments for them.

**Comment by SchrodingersGat** [comment]
@bugobliterator changes made as requested :)

**Review by bugobliterator — APPROVED** [review]

**Comment by SchrodingersGat** [comment]
Thanks @bugobliterator - once this is merged I'll submit a documentation PR :)

**Comment by SchrodingersGat** [comment]
Thanks Tridge!

**Pull request merged**